File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.rs
#[inline]
fn is_vowel(c: &char) -> bool {
    matches!(c, 'a' | 'e' | 'i' | 'o' | 'u'
                | 'A' | 'E' | 'I' | 'O' | 'U')
}

impl Solution {
    pub fn reverse_vowels(s: String) -> String {
        let mut vowels: Vec<char> = s.chars().filter(is_vowel).collect();
        vowels.reverse();
        let mut rev = vowels.into_iter();
        s.chars()
            .map(|c| if is_vowel(&c) { rev.next().unwrap() } else { c })
            .collect()
    }
}

File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.cpp
class Solution {
public:
    string reverseVowels(string s) {
        auto isVowel = [](char c) {
            return c == 'a' || c == 'e' || c == 'i' ||
                   c == 'o' || c == 'u' || c == 'A' ||
                   c == 'E' || c == 'I' || c == 'O' ||
                   c == 'U';
        };

        int left = 0;
        int right = static_cast<int>(s.size()) - 1;

        while (left < right) {
            if (!isVowel(s[left])) {
                ++left;
            } else if (!isVowel(s[right])) {
                --right;
            } else {
                swap(s[left], s[right]);
                ++left;
                --right;
            }
        }

        return s;
    }
};